src/charm_state.py

```diff
@@ -0,0 +1,201 @@
+        self._webserver_threads = webserver_threads
+        self._webserver_keepalive = webserver_keepalive
+        self._webserver_timeout = webserver_timeout
+        self._flask_config = flask_config if flask_config is not None else {}
+
+    @classmethod
+    def from_charm(cls, charm: CharmBase) -> "CharmState":
+        """Initialize a new instance of the CharmState class from the associated charm.
+
+        Args:
+            charm: The charm instance associated with this state.
+
+        Return:
+            The CharmState instance created by the provided charm.
+
+        Raises:
+            CharmConfigInvalidError: if the charm configuration is invalid.
+        """
+        keepalive = charm.config.get("webserver_keepalive")
+        timeout = charm.config.get("webserver_timeout")
+        workers = charm.config.get("webserver_workers")
+        threads = charm.config.get("webserver_threads")
+        flask_config = {
+            k.removeprefix("flask_"): v for k, v in charm.config.items() if k.startswith("flask_")
+        }
+        try:
+            valid_flask_config = FlaskConfig(**flask_config)  # type: ignore
+        except ValidationError as exc:
+            error_fields = set(
+                itertools.chain.from_iterable(error["loc"] for error in exc.errors())
+            )
+            error_field_str = " ".join(f"flask_{f}" for f in error_fields)
+            raise CharmConfigInvalidError(f"invalid configuration: {error_field_str}") from exc
+        return cls(
+            flask_config=valid_flask_config.dict(exclude_unset=True, exclude_none=True),
+            webserver_workers=int(workers) if workers is not None else None,
+            webserver_threads=int(threads) if threads is not None else None,
+            webserver_keepalive=int(keepalive) if keepalive is not None else None,
+            webserver_timeout=int(timeout) if timeout is not None else None,
+        )
+
+    @property
+    def webserver_config(self) -> WebserverConfig:
+        """Get the web server configuration file content for the charm.
+
+        Returns:
+            The web server configuration file content for the charm.
+        """
+        return WebserverConfig(
+            workers=self._webserver_workers,
+            threads=self._webserver_threads,
+            keepalive=datetime.timedelta(seconds=int(self._webserver_keepalive))
+            if self._webserver_keepalive is not None
+            else None,
+            timeout=datetime.timedelta(seconds=int(self._webserver_timeout))
+            if self._webserver_timeout is not None
+            else None,
+        )
+
+    @property
+    def flask_config(self) -> dict[str, str | int | bool]:
+        """Get the value of the flask_config charm configuration.
+
+        Returns:
+            The value of the flask_config charm configuration.
+        """
+        return self._flask_config.copy()
+
+    @property
+    def base_dir(self) -> pathlib.Path:
```

Comment:
Would these be better as constants since they are not quite "states"?

Reply:
One advantage of properties is that it's immutable and it can be define as `abstractmethod` which I am considering doing it in the future. I think it's okay to leave this as is for now.